File: app/src/main/java/com/vocaby/app/viewmodels/MyEntryViewModel.java
package com.vocaby.app.viewmodels;

import static com.vocaby.app.Constants.ITEM_PAYLOAD_KEY;

import android.app.Activity;
import android.app.Application;
import android.content.Intent;

import androidx.activity.result.ActivityResult;
import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.vocaby.app.models.ItemIntPayload;
import com.vocaby.app.models.ItemState;
import com.vocaby.app.models.ItemStringPayload;
import com.vocaby.app.data.VocabyRepository;
import com.vocaby.app.utils.Logger;
import com.vocaby.app.utils.SingleLiveEvent;

import java.util.ArrayList;
import java.util.List;

import io.reactivex.rxjava3.disposables.CompositeDisposable;

public class MyEntryViewModel extends AndroidViewModel {
    private List<String> customEntries;

    private  final VocabyRepository vocabyRepository;
    private final CompositeDisposable compositeDisposable;

    private int selectedPosition;

    private final MutableLiveData<Integer> mEntryCount;
    private final SingleLiveEvent<ItemIntPayload> mEntryBuilderResultPayload;
    private final SingleLiveEvent<List<String>> mEntries;
    private final SingleLiveEvent<ItemIntPayload> mDeleteStatus;

    public MyEntryViewModel(@NonNull Application application) {
        super(application);
        vocabyRepository = new VocabyRepository(application);
        compositeDisposable = new CompositeDisposable();

        selectedPosition = -1;

        mEntryCount = new MutableLiveData<>(0);
        mEntryBuilderResultPayload = new SingleLiveEvent<>();
        mEntries = new SingleLiveEvent<>();
        mDeleteStatus = new SingleLiveEvent<>();

        compositeDisposable.add(
                vocabyRepository.getCurrentUserId()
                .flatMap(vocabyRepository::getUserEntries)
                .subscribe(list -> {
                    customEntries = new ArrayList<>(list);
                    mEntries.setValue(customEntries);
                    mEntryCount.setValue(customEntries.size());
                }, Logger::reportError)
        );
    }

    public LiveData<List<String>> getEntries() {
        return mEntries;
    }
    public LiveData<ItemIntPayload> getEntryResultPayload() { return mEntryBuilderResultPayload; }
    public LiveData<Integer> getCustomEntryCount() {
        return mEntryCount;
    }
    public LiveData<ItemIntPayload> getDeleteStatus() { return mDeleteStatus; }

    public Intent addEntryDataToIntent(Intent intent, String entry, int position) {
        int selectedPosition;
        ItemStringPayload itemStringPayload = new ItemStringPayload(entry);
        if (position == -1) {
            selectedPosition = customEntries.indexOf(entry);
        } else {
            selectedPosition = position;
        }

        if (selectedPosition == -1) {
            itemStringPayload.setState(ItemState.ADD);
        } else {
            itemStringPayload.setState(ItemState.UPDATE);
        }

        intent.putExtra(ITEM_PAYLOAD_KEY, itemStringPayload);
        return intent;
    }

    public void handleResult(ActivityResult result) {
        if (result.getData() != null && result.getResultCode() == Activity.RESULT_OK) {
            ItemStringPayload receivedPayload =
                    result.getData().getParcelableExtra(ITEM_PAYLOAD_KEY);

            if (receivedPayload.getState() == ItemState.ADD) {
                customEntries.add(0, receivedPayload.getPayload());
            } else if (receivedPayload.getState() == ItemState.DELETE){
                if (selectedPosition != -1 ) customEntries.remove(selectedPosition);
                else customEntries.remove(receivedPayload.getPayload());
            }

            mEntryBuilderResultPayload.setValue(new ItemIntPayload(receivedPayload.getState(), selectedPosition));
            mEntryCount.setValue(customEntries.size());
        }
    }

    public void removeCustomEntry(String entry, int position) {
        compositeDisposable.add(
                vocabyRepository.deleteUserEntry(entry)
                    .subscribe(() -> {
                        customEntries.remove(position);
                        mDeleteStatus.setValue(new ItemIntPayload(ItemState.DELETE, position));
                        mEntryCount.setValue(customEntries.size());
                    }, error -> {
                        mDeleteStatus.setValue(new ItemIntPayload(ItemState.UNCHANGED, position));
                        Logger.reportError(error);
                    })
        );
    }

    public void setSelectedPosition(int selected) {
        selectedPosition = selected;
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        compositeDisposable.clear();
    }
}

File: app/src/main/java/com/vocaby/app/viewmodels/MyEntryViewModel.kt
package com.vocaby.app.viewmodels

import android.app.Activity
import android.content.Intent
import androidx.activity.result.ActivityResult
import androidx.lifecycle.*
import com.vocaby.app.Constants
import com.vocaby.app.data.VocabyRepository
import com.vocaby.app.models.customentry.UserEntry
import com.vocaby.app.payloads.ItemEntryPayload
import com.vocaby.app.payloads.ItemIntPayload
import com.vocaby.app.payloads.ItemStringPayload
import com.vocaby.app.states.ItemState
import com.vocaby.app.states.UserInputState
import com.vocaby.app.utils.Formatter
import com.vocaby.app.utils.SingleLiveEvent
import kotlinx.coroutines.launch
import java.util.*

class MyEntryViewModel(private val repository: VocabyRepository): ViewModel() {
    private val _entryCount: MutableLiveData<Int> = MutableLiveData(0)
    private val _entries: SingleLiveEvent<List<UserEntry>> = SingleLiveEvent()
    private val _entryState: SingleLiveEvent<ItemIntPayload> = SingleLiveEvent()
    private val _userInput: SingleLiveEvent<UserInputState> = SingleLiveEvent()

    private var selectedPosition: Int = -1
    private var customEntries: MutableList<UserEntry> = ArrayList()

    val entries: LiveData<List<UserEntry>>
        get() = _entries
    val customEntryCount: LiveData<Int>
        get() = _entryCount
    val entryResult: LiveData<ItemIntPayload>
        get() = _entryState
    val entryCreationState: LiveData<UserInputState>
        get() = _userInput

    init {
        // populate user entries
        initializeEntries()
    }

    fun clearEntries() {
        viewModelScope.launch {
            repository.clearUserEntries()
            customEntries = ArrayList()
            _entries.postValue(customEntries)
            _entryCount.postValue(0)
        }
    }

    fun addEntryDataToIntent(intent: Intent, entry: String, position: Int): Intent {
        val itemPayload = ItemStringPayload(entry, ItemState.ADD)

        selectedPosition = if (position == -1) {
            var pos = -1
            for ((index, userEntry) in customEntries.withIndex()) {
                if (userEntry.entry == entry) {
                    pos = index
                }
            }

            pos
        } else {
            position
        }

        if (selectedPosition != -1) {
            itemPayload.state = ItemState.UPDATE
        }

        intent.putExtra(Constants.ITEM_PAYLOAD_KEY, itemPayload)
        return intent
    }

    fun handleResult(result: ActivityResult) {
        if (result.data != null && result.resultCode == Activity.RESULT_OK) {
            val receivedPayload: ItemEntryPayload? =
                result.data!!.getParcelableExtra(Constants.ITEM_PAYLOAD_KEY)

            receivedPayload?.let {
                if (receivedPayload.state == ItemState.ADD) {
                    customEntries.add(0, receivedPayload.payload)
                } else if (receivedPayload.state == ItemState.DELETE && selectedPosition != -1) {
                    customEntries.removeAt(selectedPosition)
                } else if (receivedPayload.state == ItemState.UPDATE) {
                    customEntries.set(selectedPosition, receivedPayload.payload)
                }

                _entryState.value = ItemIntPayload(selectedPosition, receivedPayload.state)
                _entryCount.value = customEntries.size
            }
        }
    }

    fun removeCustomEntry(entry: String, position: Int) {
        viewModelScope.launch {
            repository.removeCustomEntry(entry)
            customEntries.removeAt(position)
            _entryCount.postValue(customEntries.size)
            _entryState.value = ItemIntPayload(position, ItemState.DELETE)
        }
    }

    fun createCustomEntry(entry:String) {
        if (entry.isEmpty()) {
            _userInput.value = UserInputState.EmptyInput
        } else if (Formatter.containsSpecialCharacter(entry)) {
            _userInput.value = UserInputState.InvalidInput
        } else {
            val sanitizedEntry = Formatter.cleanText(entry)
            if (sanitizedEntry.length > Constants.ENTRY_MAX_LENGTH) {
                _userInput.value = UserInputState.LongInput
            } else {
                _userInput.value = UserInputState.Valid(sanitizedEntry)
            }
        }
    }

    fun initializeEntries() {
        viewModelScope.launch {
            customEntries = repository.getUserEntries() as MutableList<UserEntry>
            _entries.postValue(customEntries)
            _entryCount.postValue(customEntries.size)
        }
    }

    fun reinitializeEntries() {
        _entries.postValue(customEntries)
        _entryCount.postValue(customEntries.size)
    }
}

class MyEntryViewModelFactory(
    private val repository: VocabyRepository
) : ViewModelProvider.Factory {
    override fun <T : ViewModel> create(modelClass: Class<T>): T {
        if (modelClass.isAssignableFrom(MyEntryViewModel::class.java)) {
            @Suppress("UNCHECKED_CAST")
            return MyEntryViewModel(repository) as T
        }
        throw IllegalArgumentException("Unknown ViewModel class")
    }
}